### ISSUE TITLE
The 48 variational parameters of QuantumLayer have identically zero gradient (circuit is diagonal + permutation only)

### ISSUE THREAD

**Issue by Kuo-TingKai** [ChiShengChen/Quantum_MNIST_Diffusion#1]
Hi, thanks for open-sourcing the code for arXiv:2504.00034 — the low-data angle is a genuinely interesting question and it is rare to get runnable code with a QML paper.

While reading the code I think I found a problem that changes how the quantum results should be interpreted, so I wanted to report it carefully with a reproduction rather than just assert it.

## Claim

**In `QuantumLayer`, the `weights` tensor has an identically zero gradient. The 48 variational parameters cannot be trained, and the layer's output does not depend on them at all.**

This affects `quantum_difussion_mnist_v7.py`, `quantum_difussion_pathmnist_v7.py` and `full_unet/quantum_diffusion_mnist_v8.py`, which all use the same ansatz.

## Why

The circuit is:

```python
for i in range(n_qubits):
    qml.RY(inputs[i], wires=i)            # encoding
for l in range(n_layers):
    for i in range(n_qubits):
        qml.RZ(weights[l, i], wires=i)    # the only trainable gates
    for i in range(n_qubits - 1):
        qml.CNOT(wires=[i, i + 1])
return [qml.expval(qml.PauliZ(i)) for i in range(n_qubits)]
```

Every gate applied *after* the encoding layer is either

* **diagonal in the computational basis** (`RZ`, which only multiplies amplitudes by unit-modulus phases), or
* a **computational-basis permutation** (`CNOT`, which only relabels basis states).

Neither operation can change the set of squared amplitudes $\{|c_b|^2\}$ — the phases cancel in $|c_b|^2$, and the permutation merely reorders them. Since $\langle Z_i \rangle = \sum_b |c_b|^2 (-1)^{b_i}$ depends *only* on those squared amplitudes and on the fixed CNOT permutation, we get

$$\frac{\partial \langle Z_i\rangle}{\partial w_{l,j}} = 0 \quad \text{for all } i, l, j .$$

This is the general statement: *a circuit built exclusively from computational-basis-diagonal gates and computational-basis permutations cannot change computational-basis measurement statistics.* To make `RZ` matter you need at least one non-diagonal parameterized rotation after entanglement, or a measurement in a rotated basis.

## Closed form

Because the CNOT ladder implements a prefix-XOR on the basis labels and the encoded state is a product state with $\mathbb{E}[(-1)^{b_j}] = \cos(x_j)$, the whole 16-qubit layer collapses to

$$\langle Z_i \rangle = \prod_{j \in S_i} \cos(x_j)$$

where $S_i$ is the (weight-independent) subset of input wires XORed into wire $i$ after `n_layers` ladders. For `n_qubits=16, n_layers=3` the subset sizes are `[1,2,2,2,3,4,4,4,5,6,6,6,7,8,8,8]`.

So the "quantum attention layer" is a **fixed, parameter-free function of its input that can be evaluated in O(n) on a CPU**. Note in particular that $\langle Z_0 \rangle = \cos(x_0)$ exactly — wire 0 is only ever a CNOT *control*, so it carries no information from any other wire and no information from any weight.

## Reproduction

Self-contained, numpy only (no PennyLane needed — it is a direct statevector simulation of the circuit as written):

```python
import numpy as np
rng = np.random.default_rng(0)
N, L = 16, 3

def apply_1q(psi, U, q):
    psi = psi.reshape((2**q, 2, 2**(N-q-1)))
    return np.einsum('ij,ajb->aib', U, psi).reshape(-1)

def apply_cnot(psi, c, t):
    psi = psi.reshape([2]*N).copy()
    idx = [slice(None)]*N; idx[c] = 1
    psi[tuple(idx)] = np.flip(psi[tuple(idx)], axis=(t if t < c else t-1))
    return psi.reshape(-1)

RY = lambda a: np.array([[np.cos(a/2), -np.sin(a/2)], [np.sin(a/2), np.cos(a/2)]], dtype=complex)
RZ = lambda a: np.array([[np.exp(-1j*a/2), 0], [0, np.exp(1j*a/2)]], dtype=complex)

def run(inputs, weights):
    psi = np.zeros(2**N, dtype=complex); psi[0] = 1.0
    for i in range(N): psi = apply_1q(psi, RY(inputs[i]), i)
    for l in range(L):
        for i in range(N): psi = apply_1q(psi, RZ(weights[l, i]), i)
        for i in range(N-1): psi = apply_cnot(psi, i, i+1)
    p = (np.abs(psi)**2).reshape([2]*N)
    return np.array([p.sum(axis=tuple(j for j in range(N) if j != i)) @ np.array([1, -1])
                     for i in range(N)])

x  = rng.uniform(-np.pi, np.pi, N)
w1 = rng.uniform(-np.pi, np.pi, (L, N))
w2 = rng.uniform(-np.pi, np.pi, (L, N))
print(np.abs(run(x, w1) - run(x, w2)).max())   # -> 1.4e-15  (i.e. exactly 0)
```

Output on my machine: `1.443e-15`, i.e. two completely different weight tensors give bit-for-bit identical outputs up to float64 round-off. I also verified the closed form above reproduces the statevector result to ~1e-16 on all 16 wires.

## Consequence

Combined with #2 (the `.detach()` call), in `v7` the only part of the quantum branch that receives any gradient at all is `output_proj` (`nn.Linear(16, 128)`). The reported FID/SSIM differences between the "quantum" and "classical" configurations therefore cannot be attributed to a trained quantum circuit — they come from a fixed cosine-product feature map plus one extra trainable linear layer that the classical baseline does not have (see #3).

## Suggested fix

Any of these breaks the diagonal+permutation structure and restores a non-zero gradient:

1. Use `qml.StronglyEntanglingLayers` (RZ-RY-RZ per wire + ring CNOTs) — the standard choice, and it lets you cite a well-characterized ansatz.
2. Keep the hand-rolled circuit but add a non-diagonal rotation per layer, e.g. `qml.RY(weights[l, i, 1], wires=i)` after the `RZ`.
3. Measure in a rotated basis, e.g. `qml.expval(qml.PauliX(i))`, so the `RZ` phases become observable.

Option 1 or 2 would be the minimal change; either way the MNIST and PathMNIST numbers would need to be regenerated.

Happy to open a PR with the fix + a unit test that asserts `weights.grad` is non-zero, if that would be useful.
